### PR TITLE
Make e2e.go give us JUnit results.

### PR DESCRIPTION
**Pull request by spxtr**
Will assign to @fejta when it's ready to go. https://github.com/kubernetes/test-infra/issues/76

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33240)

### CODE REVIEW

**Comment by spxtr** [comment]
cc @maisem

**Comment by maisem** [comment]
This should fix the GCE upgrade tests as well.
cc @pwittrock @mikedanese @fabioy

**Comment by mikedanese** [comment]
Or break them once they start testing the right thing :)

**Comment by spxtr** [comment]
OK, assigned to @fejta for review. Don't merge until we drop that testing commit.

**Comment by spxtr** [comment]
Successful output looks like this:
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuite failures="0" tests="5" time="827.137401602">
    <testcase classname="e2e.go" name="TearDown" time="43.999470684"></testcase>
    <testcase classname="e2e.go" name="Up" time="560.392844715"></testcase>
    <testcase classname="e2e.go" name="kubectl version" time="0.16694108300000002"></testcase>
    <testcase classname="e2e.go" name="Test" time="117.80508784"></testcase>
    <testcase classname="e2e.go" name="TearDown" time="104.772889835"></testcase>
</testsuite>
```

File: hack/e2e-internal/e2e-up.sh [hunk 1]
```diff
@@ -29,6 +29,9 @@ source "${KUBE_ROOT}/cluster/kube-util.sh"
 
 prepare-e2e
 
+echo "MUAHAHA: DO NOT MERGE"
```
Comment (fejta) [inline hack/e2e-internal/e2e-up.sh]: Please resolve ;-)

File: hack/e2e.go [hunk 2]
```diff
@@ -18,46 +18,39 @@ limitations under the License.
...
 	testArgs             = flag.String("test_args", "", "Space-separated list of arguments to pass to Ginkgo test runner.")
 	test                 = flag.Bool("test", false, "Run Ginkgo tests.")
 	up                   = flag.Bool("up", false, "If true, start the the e2e cluster. If cluster is already up, recreate it.")
 	upgradeArgs          = flag.String("upgrade_args", "", "If set, run upgrade tests before other tests")
 	verbose              = flag.Bool("v", false, "If true, print all command output.")
-)
 
-const (
-	minNodeCount = 2
+	push   = flag.Bool("push", false, "Deprecated. Does nothing.")
```
Comment (fejta) [inline hack/e2e.go]: deprecatedPush? Here and below?

File: hack/e2e.go [hunk 3]
```diff
@@ -67,125 +60,198 @@ func appendError(errs []error, err error) []error {
...
+		ClassName: "e2e.go",
+		Time:      duration.Seconds(),
+	}
+	if err != nil {
+		c.Failure = err.Error()
+		suite.Failures++
+	}
+	suite.Cases = append(suite.Cases, c)
+	suite.Tests++
+	return err
```
Comment (fejta) [inline hack/e2e.go]: Doesn't this mean that every err stack trace will be at e2e.go:112?

**Comment by spxtr** [comment]
Failed XML looks like https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/33240/kubernetes-pull-build-test-e2e-gke/13756/
```
<?xml version="1.0" encoding="UTF-8"?>
<testsuite failures="1" tests="2" time="41.105718882">
    <testcase classname="e2e.go" name="TearDown" time="41.089728814"></testcase>
    <testcase classname="e2e.go" name="Up" time="0.015956259">
        <failure>error running up: exit status 1</failure>
    </testcase>
</testsuite>
```
It would be sweet to have better failure messages, but this seems like a good start.

**Review by fejta — APPROVED** [review]

File: hack/e2e-internal/e2e-up.sh (hunk 1 above)
Comment (spxtr) [inline hack/e2e-internal/e2e-up.sh]: Done.

File: hack/e2e.go (hunk 3 above)
Comment (spxtr) [inline hack/e2e.go]: Not sure what's going to cause a stack trace. Errors look like this:
```
2016/09/22 18:31:19 e2e.go:468: Step 'up' finished in 15.909371ms
2016/09/22 18:31:19 e2e.go:133: Saved XML output to /workspace/_artifacts/junit_runner.xml.
2016/09/22 18:31:19 e2e.go:145: Something went wrong: starting e2e cluster: error running up: exit status 1
```

File: hack/e2e.go (hunk 2 above)
Comment (spxtr) [inline hack/e2e.go]: Done.

File: hack/e2e.go (hunk 3 above)
Comment (fejta) [inline hack/e2e.go]: Got it because you're just returning the error (that's what you cannot change, creation of the error)

**Review by fejta — APPROVED** [review]

**Pull request merged**